### PR TITLE
Fix mailbox flooding with Poll messages in KafkaConsumerActor (#236)

### PR DESCRIPTION
**Pull request by iravid**
If the execution time of `poll()` is longer than the configured
`pollInterval`, the actor's mailbox could be flooded with `Poll`
messages, starving out other messages.

This commit changes the scheduling mechanism to use `scheduleOnce`: a
`Poll` message will be scheduled in `preStart()`, and another will be
scheduled once the `Poll` message was processed.

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @iravid,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 1]
```diff
@@ -160,11 +161,14 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
...
     super.preStart()
+
     consumer = settings.createKafkaConsumer()
+    currentPollTask = schedulePollTask()
   }
 
   override def postStop(): Unit = {
-    pollTask.cancel()
+    if (currentPollTask != null)
+      currentPollTask.cancel()
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: The problem with this solution is that there is a small (but existing) chance that there is already a `Poll` message in flight in the actor's mailbox when the actor is **restarted**, and that can double the frequency of the polls.

One way to solve that is to include a `target: KafkaConsumer` in the `Poll` message and when receiving `Poll` ignore any `target != this`.

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 2]
```diff
@@ -38,7 +38,7 @@ object KafkaConsumerActor {
     final case class Messages[K, V](requestId: Int, messages: Iterator[ConsumerRecord[K, V]])
     final case class Committed(offsets: Map[TopicPartition, OffsetAndMetadata])
     //internal
-    private[KafkaConsumerActor] case object Poll extends DeadLetterSuppression
+    private[KafkaConsumerActor] case class Poll(target: ActorRef) extends DeadLetterSuppression
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: always add `final` to `case class`

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 1 above)
Comment (iravid) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Nice :-) Hadn't thought of that. I've implemented your suggested solution. `this` would be different across restarts as opposed to `self`, right?

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 3]
```diff
@@ -114,9 +114,16 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
       consumer.subscribe(topics.toList.asJava, new WrappedAutoPausedListener(consumer, listener))
     case SubscribePattern(pattern, listener) =>
       consumer.subscribe(Pattern.compile(pattern), new WrappedAutoPausedListener(consumer, listener))
-    case Poll =>
-      poll()
-      currentPollTask = schedulePollTask()
+    case Poll(target) =>
+      if (target == self) {
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: no, it must be an instance check, the `ActorRef` will be the same after restart.
`target: KafkaConsumerActor` and `target == this`

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 4]
```diff
@@ -180,7 +194,7 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
   }
 
   def schedulePollTask(): Cancellable =
-    context.system.scheduler.scheduleOnce(pollInterval(), self, Poll)(context.dispatcher)
+    context.system.scheduler.scheduleOnce(pollInterval(), self, Poll(self))(context.dispatcher)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: to avoid creating a new `Poll` instance all the time:
`val pollMsg = Poll(this)`

**Comment by iravid** [comment]
Let me know if you'd like me to squash after review is done

**Comment by patriknw** [comment]
we can squash from the github ui before merging, thanks

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 1 above)
Comment (iravid) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Ah, wait, that cannot work - `this` isn't an ActorRef. Is there a way to implement this when comparing `self`?

Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: yes, you want to check that the actor instance (this) is the right one

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 4 above)
Comment (iravid) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Added above pollTimeout

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 3 above)
Comment (iravid) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Understood now :-)

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 2 above)
Comment (iravid) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: 👍

**Comment by iravid** [comment]
Thanks for the review

**Pull request merged**